### ISSUE TITLE
How to prevent buffering while Player is paused

### ISSUE THREAD

**Issue by y20k** [androidx/media#146]
I am currently porting a streaming radio app over to use *media3*.

The expected behavior for a radio app regarding start/stop of playback is my opinion the following:

1. user starts playback => app starts playing live radio
2. user stops/pauses playback => app stops playing
3. user starts playback again => app starts playing live radio, content that the user missed while the playback was stopped is skipped, radio is always *live*

Basically I think a radio app should behave like an old school FM radio. It is always live. Playback should not be behind. One should not be able to rewind / fast forward.

**Solution 1: use `Player.stop()`**
After the user tapped on Stop the apps stops buffering immediately. Which is good. But due to the automatic foreground handling of the app (see also #45) the notification is dismissed immediately. That is not good. Users expect to be able to start playback again after hitting Stop or Pause - at least for a couple of minutes.

**Solution 2: use `Player.pause()`**
The Player buffers the live stream and resumes playback from the position, it has been paused. The playback is behind. The radio is not *live* anymore. I can change that behavior and `seekTo(0)` to resume *live* playback. My problem is the buffering. That produce unnecessary network traffic.

I now wonder how I can proceed. **Is there a way to tell the Player to stop buffering when it has been paused?**

PS. *media3* is a lot easier to use and I need way less code to the things I need it to do than *support.v4.media*. Thank you for the great work here.

**Comment by christosts** [comment]
Maybe try building the player with a custom DefaultLoadControl whose shouldContinueLoading() returns false when you pause the player.

I have not tested this but it might work.

**Comment by y20k** [comment]
Hi @christosts 
I have tried your suggestion. It seems that `shouldContinueLoading()` is only called when playback starts. It is not called when playback is stopped or paused.

**Comment by christosts** [comment]
`shouldContinueLoading()` will be called while the player is paused, if the player thinks it can download more content, e..g it will *not* be called if the player has already buffered until the end of a file. Maybe you're testing with a file is already buffered when you pause? I quickly tested the following custom load control

```java
public class MyLoadControl extends DefaultLoadControl implements Player.Listener {
  private boolean paused;

  @Override
  public boolean shouldContinueLoading(long playbackPositionUs, long bufferedDurationUs,
      float playbackSpeed) {
    if (paused) {
      return false;
    }
    return r = super.shouldContinueLoading(playbackPositionUs, bufferedDurationUs, playbackSpeed);
  }

  @Override
  public void onPlayWhenReadyChanged(boolean playWhenReady, int reason) {
    paused = !playWhenReady;
  }
}
```

and set it in the demo-session app's `PlaybackService.initializeSessionAndPlayer()`
```kotlin
var loadControl = MyLoadControl()
player =
        ExoPlayer.Builder(this)
        .setAudioAttributes(AudioAttributes.DEFAULT, /* handleAudioFocus= */ true)
        .setLoadControl(loadControl)
        .build()
player.addListener(loadControl)
```

and I see that downloading media stops when the player is paused.

**Comment by y20k** [comment]
Hi @christosts 

my implementation of for a custom `LoadControl` was pretty similar to what you suggested. Although I initially used `onIsPlayingChanged` in my tests to determine and store the paused state. In both cases my observations were similar.

1) When the Player is paused `shouldContinueLoading` is not being called for several minutes. (I mistakenly thought I was never called, sorry)

2) When `shouldContinueLoading` is called it returns `false`as intended

3) But even after it returned `false` network continues unchanged

See also the attached screenshot. The streaming URL I used was: http://fm939.wnyc.org/wnycfm.aac I will try to reproduce the issue using the session demo app.

![onIsPlayingChanged](https://user-images.githubusercontent.com/9103935/184147302-86c19014-bb9c-424b-9526-9e5fbefeaab3.png)

**Comment by y20k** [comment]
When I implement the `CustomLoadControl` I used in my app for testing purposes in the Session Demo App, I get the results described above. 

> 1. When the Player is paused shouldContinueLoading is not being called for several minutes
> 2. When shouldContinueLoading is called it returns false as intended
> 3. But even after it returned false network traffic continues unchanged

Here are the changes I made in detail https://github.com/androidx/media/compare/release...y20k:media3:release

**Comment by christosts** [comment]
This is audio file is loaded with the `ProgressiveMediaSource`, which by default loads [1 MB](https://github.com/androidx/media/blob/1.0.0-beta02/libraries/exoplayer/src/main/java/androidx/media3/exoplayer/source/ProgressiveMediaSource.java#L235) of data before checking again if it should continue loading. To stop loading sooner, you can set a lower setting by setting a custom media source factory on the player

```kotlin
var loadControl = MyLoadControl()
    player =
      ExoPlayer.Builder(this)
        .setLoadControl(loadControl)
        .setMediaSourceFactory(
          ProgressiveMediaSource.Factory(DefaultDataSource.Factory(context))
            // load 64KB at a time before checking with the load control
            .setContinueLoadingCheckIntervalBytes(64 * 1024)
        )
        .build()
    player.addListener(loadControl)
```

Note, with the above snippet, you can only play progressive media

**Comment by y20k** [comment]
Hi @christosts 

Again, thank you for taking your time here!! I totally would understand if this issue was not important enough to continue to investigate it. Anyway. My findings here are:

Following you suggestion I [added](https://github.com/y20k/media3/commit/2abe1b72dc400292de4906311fb998ab99886e26) a custom `MediaSourceFactory`. Now there is no delay anymore. `shouldContinueLoading` is called immediately. With the custom `MediaSourceFactory` there was a three minute delay. 

But nothing changed in the general observation: the Demo app keeps loading data for several minutes after `shouldContinueLoading` returned `false`.

![customloadcontrol-and-custommediasource](https://user-images.githubusercontent.com/9103935/184347233-b1b09da1-01b7-45a1-ae7a-ce27b2233e7e.png)

In this screenshot you can see: 

- I paused playback after ~**15 seconds** (Logcat: time span "onPlayWhenReadyChanged - paused = false" to "onPlayWhenReadyChanged - paused = true"
- data continues to be loaded - actually for **4 minutes** @ ~5 KB/s (see the second screenshot) The data rate does not change after playback has been paused.

![4min-networktraffic](https://user-images.githubusercontent.com/9103935/184348670-391c8e19-c22a-440b-8d74-1b86c67c7d97.png)

**Comment by christosts** [comment]
When I tested this, I did not see any calls to load content beyond the 64KB interval. I'm not sure what else might be triggering network traffic. You can monitor if the player is loading more than the 64KB "interval" by adding a break point in `DefaultHttpDataSource.read()`. This method will be called a few times until "interval bytes" have been downloaded since the last time `shouldContinueLoading` returned true.

**Comment by y20k** [comment]
Hi @christosts

> `DefaultHttpDataSource.read()`. This method will be called a few times until "interval bytes" have been downloaded since the last time `shouldContinueLoading` returned true.

I can confirm that :+1:. I [added](https://github.com/androidx/media/commit/b58ef84f2cdc6cc528f65dc38aa58a3838f62d3a) a Log command at `DefaultHttpDataSource.readInternal()`. `DefaultHttpDataSource.read()` is called repeatedly for only 10 seconds after `shouldContinueLoading` returned false.

But I still see in that the Demo app produces Network traffic for 3 1/2 minutes after playback has been paused.

![network-traffic](https://user-images.githubusercontent.com/9103935/184389723-5cbb9afc-5a83-4bac-b24e-2970fcf53e55.png)

**Comment by christosts** [comment]
We suspect the network traffic is coming from the network library. While paused for progressive media, we do not close the connection, so it could be that the underlying network stack continues reading the file until it fills some internal buffer. It's likely that what you are seeing is the network stack filling up ~1MB.

Things you could try:
- `Player.stop` will close the connection, though you mentioned you want to avoid it because it will hide the notification. You could implement notifications on your own
- Try a different network stack ([ExoPlayer documentation](https://exoplayer.dev/network-stacks.html)) in case that changes what you observe.

Last one, instead of overriding `shouldContinueLoading`, you can probably achieve similar behavior if you limit the `DefaultLoadControl`'s `maxBufferMs` to a small value (5 seconds?).

I'll close this now as I think the question originally posted has been answered.

**Comment by y20k** [comment]
Thank you. My original question is answered as well as the follow-up questions.
I‘ve learned a lot and I hope your analysis and explanation will benefit other developers, too.

**Comment by y20k** [comment]
Just for the sake of completeness. 

* I tried using an `OkHttpDataSource` to check, if a different network stack would change the observed behavior. 
* I also tried to set `maxBufferMs` to 5000ms.

The results - both individually and combined - are the same: The app continues to produce network traffic several minutes after the Player was paused.

I am now need to decide (a) if that is acceptable, since the amount of data transferred is not that big, of (b) if a implement my own notification and use `Player.stop()`. But that decision is not of interest for this issue. :-)
